src/main/scala/net/snowflake/spark/snowflake/io/SnowflakeResultSetRDD.scala

```diff
@@ -125,9 +127,19 @@ case class ResultIterator[T: ClassTag](
     }
   }
 
+  context.addTaskCompletionListener[Unit](_ => closeResultSet())
+
   val isIR: Boolean = isInternalRow[T]
   val mapper: ObjectMapper = new ObjectMapper()
   var currentRowNotConsumedYet: Boolean = false
+  var resultSetIsClosed: Boolean = false
+
+  private def closeResultSet(): Unit = {
+    if (!resultSetIsClosed) {
+      data.close()
+      resultSetIsClosed = true
+    }
```

Comment:
```
      resultSetIsClosed = true
      data.close()
```

Reply:
@JoshRosen Appreciated so much for your contributions! Left some comments.

Reply:
@sfc-gh-mrui, I've pushed a commit addressing this round of comments.